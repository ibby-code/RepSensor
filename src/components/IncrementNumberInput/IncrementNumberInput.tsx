import React, { FC, useState } from 'react';
import { Button, Input, XStack } from 'tamagui';


interface IncrementNumberInputProps {
  label: string;
  value?: number;
  delta: number;
  onChange?: (value: number) => void;
  // Defaults to False.
  disableFloats?: boolean;
}

/** 
 * Displays an input with the ability to increment/decrement the value.
 * Floats are limited to two decimal places.
 * Example drawing: <-> [Enter a value] <+>
 */
const IncrementNumberInput: FC<IncrementNumberInputProps> =
  ({ label, value = "", delta, onChange, disableFloats = false }) => {
    const [valueStr, setValueStr] = useState(value.toString())
    const setValueNum = (val: string, change: number) => {
      if (!val) val = '0';
      let valNum = disableFloats ? parseInt(val) : parseFloat(val);
      if (!isNaN(valNum)) {
        // Rond to two decimal places
        valNum = Math.round(valNum * 100) / 100;
        const newVal = Math.max(valNum + change, 0);
        setValueStr(newVal.toString());
        if (onChange) onChange(newVal);
      }
    }
    return (
      <XStack>
        <Button onPress={() => setValueNum(valueStr, -delta)}>-</Button>
        <Input
          placeholder={label}
          onChangeText={(text) => setValueStr(text)}
          onBlur={() => setValueNum(valueStr, 0)}
          keyboardType="numeric"
          value={valueStr} />
        <Button onPress={() => setValueNum(valueStr, delta)}>+</Button>
      </XStack>
    )
  };

export default IncrementNumberInput;
