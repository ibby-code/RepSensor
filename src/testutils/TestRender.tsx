import React from 'react';
import { getDefaultTamaguiConfig } from '@tamagui/config-default'
import { TamaguiProvider, createTamagui } from 'tamagui';
import { render as renderNative} from '@testing-library/react-native';

const conf = createTamagui(getDefaultTamaguiConfig());

export function render(jsx: React.JSX.Element) {
    return renderNative(
        <TamaguiProvider config={conf}>
            {jsx}
        </TamaguiProvider>
    )
}
