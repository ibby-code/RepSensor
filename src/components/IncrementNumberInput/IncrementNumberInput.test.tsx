import React from 'react';
import IncrementNumberInput from './IncrementNumberInput';
import { fireEvent, userEvent, waitFor } from '@testing-library/react-native';
import { render } from '../../testutils/TestRender';
import { GetAllByQuery, GetByQuery } from '@testing-library/react-native/build/queries/make-queries';
import { ByRoleOptions, ByRoleMatcher } from '@testing-library/react-native/build/queries/role';
import { TextMatch, TextMatchOptions } from '@testing-library/react-native/build/matches';

describe('<IncrementNumberInput />', () => {
  test('renders correctly', () => {
    const tree = render(
      <IncrementNumberInput label="Enter..." delta={1} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('sets the input value', () => {
    const placeholder = "Enter..."
    const value = 25
    const { getByPlaceholderText } = render(
      <IncrementNumberInput label={placeholder} value={value} delta={1} />
    );

    const input = getByPlaceholderText(placeholder);
    expect(input.props["value"]).toEqual(value.toString());
  });

  test('triggers onChange', () => {
    const mockCallback = jest.fn(() => 5);
    const { getByText } = render(
      <IncrementNumberInput label="Enter..."
        value={0} delta={1} onChange={() => mockCallback()} />
    );

    const decreaseButton = getByText("-");
    fireEvent.press(decreaseButton);

    expect(mockCallback).toHaveBeenCalledTimes(1)
  });

  describe('button clicks', () => {
    const placeholderText = 'Enter...';
    let getByText: GetByQuery<TextMatch, TextMatchOptions>;
    let getByPlaceholderText: GetByQuery<TextMatch, TextMatchOptions>;
    beforeEach(() => {
      const value = 25;
      const result = render(
        <IncrementNumberInput label={placeholderText} value={value} delta={5} />
      );
      getByText = result.getByText;
      getByPlaceholderText = result.getByPlaceholderText
    });

    test('increments value', () => {
      const increaseButton = getByText("+")

      fireEvent.press(increaseButton);

      const input = getByPlaceholderText(placeholderText);
      expect(input.props["value"]).toEqual("30");
    });

    test('decrements value', () => {
      const decreaseButton = getByText("-")

      fireEvent.press(decreaseButton);

      const input = getByPlaceholderText(placeholderText);
      expect(input.props["value"]).toEqual("20");
    });

    test('does not allow negative values', () => {
      const decreaseButton = getByText("-")

      fireEvent.press(decreaseButton);
      fireEvent.press(decreaseButton);
      fireEvent.press(decreaseButton);
      fireEvent.press(decreaseButton);
      fireEvent.press(decreaseButton);
      fireEvent.press(decreaseButton);

      const input = getByPlaceholderText(placeholderText);
      expect(input.props["value"]).toEqual("0");
    });
  });

  describe('floats', () => {
    const placeholderText = 'Enter...';

    test('allowed by default', async () => {
      const value = 25;
      const { getByPlaceholderText } = render(
        <IncrementNumberInput label={placeholderText} value={value} delta={5} />
      );

      const newValue = '3.27';
      userEvent.setup();
      const input = getByPlaceholderText(placeholderText);

      await userEvent.clear(input);
      await userEvent.type(input, newValue);

      await waitFor(() =>
        expect(input.props["value"]).toEqual(newValue)
      );
    });

    test('are converted on blur', async () => {
      const value = 25;
      const { getByPlaceholderText, getByText } = render(
        <IncrementNumberInput label={placeholderText} value={value} delta={5} disableFloats={true} />
      );
      const newValue = '3.27';
      userEvent.setup();
      const input = getByPlaceholderText(placeholderText);

      await userEvent.clear(input);
      await userEvent.type(input, newValue);
      const button = getByText("+");
      fireEvent.press(button)

      await waitFor(() =>
        expect(input.props["value"]).toEqual("8")
      );
    });
  });
});