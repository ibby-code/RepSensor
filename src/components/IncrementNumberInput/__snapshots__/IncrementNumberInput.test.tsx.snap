// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncrementNumberInput /> renders correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <View
    focusVisibleStyle={{}}
    focusable={true}
    role="button"
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 1,
        "borderRightColor": "transparent",
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "height": 44,
        "justifyContent": "center",
        "paddingLeft": 18,
        "paddingRight": 18,
      }
    }
  >
    <Text
      lineBreakMode="clip"
      numberOfLines={1}
      style={
        {
          "color": "#000",
          "flexGrow": 0,
          "flexShrink": 1,
          "fontFamily": "System",
          "userSelect": "none",
        }
      }
      suppressHighlighting={true}
    >
      -
    </Text>
  </View>
  <TextInput
    focusVisibleStyle={{}}
    focusable={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Enter..."
    style={
      {
        "backgroundColor": "#fff",
        "borderBottomWidth": 1,
        "borderLeftWidth": 1,
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "color": "#000",
        "fontFamily": "System",
        "height": 44,
        "minWidth": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
    value=""
  />
  <View
    focusVisibleStyle={{}}
    focusable={true}
    role="button"
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 1,
        "borderRightColor": "transparent",
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "height": 44,
        "justifyContent": "center",
        "paddingLeft": 18,
        "paddingRight": 18,
      }
    }
  >
    <Text
      lineBreakMode="clip"
      numberOfLines={1}
      style={
        {
          "color": "#000",
          "flexGrow": 0,
          "flexShrink": 1,
          "fontFamily": "System",
          "userSelect": "none",
        }
      }
      suppressHighlighting={true}
    >
      +
    </Text>
  </View>
</View>
`;
